converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/Observation/ObservationUpsertHandler.cs

```diff
@@ -0,0 +1,120 @@
+        }
+
+        public async Task<FhirTransactionRequestEntry> BuildAsync(FhirTransactionContext context, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(context, nameof(context));
+            EnsureArg.IsNotNull(context.ChangeFeedEntry, nameof(context.ChangeFeedEntry));
+            EnsureArg.IsNotNull(context.Request, nameof(context.Request));
+
+            IResourceId patientId = context.Request.Patient.ResourceId;
+            IResourceId imagingStudyId = context.Request.ImagingStudy.ResourceId;
+            ChangeFeedEntry changeFeedEntry = context.ChangeFeedEntry;
+
+            // Parse all observations out of the dataset
+            ParsedObservation parsedObservations = ObservationParser.CreateObservations(
+                changeFeedEntry.Metadata,
+                patientId.ToResourceReference(),
+                imagingStudyId.ToResourceReference()
+            );
+            Collection<Observation> doseSummaries = parsedObservations.DoseSummaries;
+            Collection<Observation> irradiationEvents = parsedObservations.IrradiationEvents;
+
+            if (!doseSummaries.Any())
+                return null;
+
+            // operate on the pre-condition that their _should_ only be one DoseSummary per ImagingStudy.
+            // If multiple observations are found; just use the first one.
+            Observation doseSummary = doseSummaries[0];
+
+            // Filter out existing observations
+            Identifier imagingStudyIdentifier = imagingStudyId.ToResourceReference().Identifier;
+            IEnumerable<Observation> existingDoseSummariesAsync = imagingStudyIdentifier != null
+                ? await _fhirService
+                    .RetrieveObservationsAsync(
+                        imagingStudyId.ToResourceReference().Identifier,
+                        cancellationToken)
+                : new List<Observation>();
+
+            var existingDoseSummaries =
+                existingDoseSummariesAsync.ToList();
+            var isExisting = existingDoseSummaries.Any();
+
+            FhirTransactionRequestMode method = isExisting
+                ? FhirTransactionRequestMode.Create
+                : FhirTransactionRequestMode.Update;
+
+            Bundle.RequestComponent request = isExisting
+                ? new Bundle.RequestComponent()
+                {
+                    Method = Bundle.HTTPVerb.PUT,
+                    Url = $"{ResourceType.Observation.GetLiteral()}/{existingDoseSummaries[0].Id}"
+                }
+                : new Bundle.RequestComponent()
+                {
+                    Method = Bundle.HTTPVerb.POST,
+                    Url = ResourceType.Observation.GetLiteral()
+                };
+
+            FhirTransactionRequestMode transactionRequestMode = isExisting
+                ? FhirTransactionRequestMode.Update
+                : FhirTransactionRequestMode.Create;
+
+            IResourceId resourceId = isExisting
```

Comment:
Is it possible to check isExisting just once and then compute FhirtransactionrequestMode, resourceId and bundlecomponent request calculation in a same  block instead?